docs/DeferredSignals.md

```diff
@@ -0,0 +1,271 @@
+
+This is the tricky memory page where we determine if we have any deferred signals to process. I say it's tricky because after FEX leaves an
+"uninterruptible" code section, it will decrement the reference counter, and then try to store in to the first byte of this page. In the case that no
+signal has been deferred, this memory store will progress without issue, consuming only two instructions in the process.
+
+In the case that a signal **HAS** been deferred, then the permissions on this page will be set to `PROT_NONE` and this memory store will cause a
+SIGSEGV. Then FEX's signal handler will check to see if the access was this special page and start the deferred signal mechanisms. This means that a
+deferred signal check is only five instructions in total.
+
+#### Example ARM64 JIT code for uninterruptible region
+```asm
+  ; Increment the reference counter.
+  ldr x0, [x28, #(offsetof(CPUState, DeferredReferenceCounter))]
+  add x0, x0, #1
+  str x0, [x28, #(offsetof(CPUState, DeferredReferenceCounter))]
+
+  ; Do the uninterruptible code section here.
+  <...>
+
+  ; Now decrement the reference counter.
+  ldr x0, [x28, #(offsetof(CPUState, DeferredReferenceCounter))]
+  sub x0, x0, #1
+  str x0, [x28, #(offsetof(CPUState, DeferredReferenceCounter))]
+
+  ; Now the magic memory access to check for any deferred signals.
+  ; Load the page pointer from the CPUState
+  ldr x0, [x28, #(offsetof(CPUState, DeferredSignalHandlerPagePtr))]
+  ; Just store zero. (1 cycle plus no dependencies on a register. Super fast!)
+  ; Will store fine with no deferred signal, or SIGSEGV if there was one!
+  str xzr, [x0]
+```
+
+### Deferred signal handling
+In the case that FEX has received a signal, FEX's signal handler will first check to see if that thread's reference counter is zero or not.
+
+#### Reference counter is zero
+This is the easy case, just handle the signal as normal.
+
+#### Reference counter is not zero
+The signal handler now knows that FEX is in an uninterruptible code section. We check the signal to see if it is a synchronous signal or not.
+- If the signal is synchronous then we need to handle it as normal, because this is a hardware signal that we can't defer.
+- If it is an async signal (from tgkill, sigqueue, or something else) then we will start the deferring process.
+
+The deferring process starts with storing the kernel `siginfo_t` to a thread local array so we can restore it later.
+We then modify the permissions on the thread local `DeferredSignalHandlerPagePtr` to be `PROT_NONE`.
+We then immediately return from the signal handler so that FEX can resume its "uninterruptible" code section without breaking anything.
+Once the "uninterruptible" code section finishes, FEX will intentionally trigger a SIGSEGV by storing to the page.
+
+This will trigger a jump to FEX's SIGSEGV handler, where FEX will process the signal as if it was the previously deferred signal.
+the previous signal that was deferred.
```

Comment:
```suggestion
This will trigger a jump to FEX's SIGSEGV handler, where FEX restores the previously saved `siginfo_t` and processes the signal as if it was the previously deferred signal.
```